core/src/main/java/hudson/slaves/SlaveComputer.java

```diff
@@ -482,7 +481,15 @@ public void onClosed(Channel c, IOException cause) {
                     taskListener.getLogger().println("Connection terminated");
                 }
                 closeChannel();
-                launcher.afterDisconnect(SlaveComputer.this, taskListener);
+                try {
+                    launcher.afterDisconnect(SlaveComputer.this, taskListener);
+                } catch (Throwable t) {
+                    LogRecord lr = new LogRecord(Level.SEVERE,
+                            "Launcher {0}'s afterDisconnect method propagated an exception when {1}'s connection was closed: {2}");
+                    lr.setThrown(t);
+                    lr.setParameters(new Object[]{launcher, SlaveComputer.this, t.getMessage()});
```

Comment:
Code becomes complex when ancient classes/methods not removed for 10 years and library forked with single patches, even java itself already cutting compatibility. 
```
logger.log(Level.SEVERE, "Launcher {0} afterDisconnect method failed: \" {1} \". {2} connection was closed", \n
                                           new Object[] {launcher,  t.getMessage(),  SlaveComputer.this});
```
And not sure that `SlaveComputer.this` is good description.